File: CustomDecoder/results/CustomDecoder.py
class XorByte:
    def __init__(self, xor_value):
        self.xor_value = xor_value

    def decode(self, src):
        dest = []
        
        i = 0
        while i < len(src):
            dest.append(src[i] ^ self.xor_value)
            i += 1
        
        return dest

class Base64:
    def decode(self, src):
        dest = []
        
        # 4 base64 chars => 3 bytes
        i = 0
        while i < len(src):
            ch0 = self.decode_char(src[i])
            ch1 = self.decode_char(src[i + 1])
            ch2 = self.decode_char(src[i + 2])
            ch3 = self.decode_char(src[i + 3])
            trinity = (ch0 << 18) + (ch1 << 12) + (ch2 << 6) + (ch3)
            dest.append(trinity >> 16)
            dest.append((trinity >> 8) & 0xff)
            dest.append(trinity & 0xff)
            i += 4
        
        return dest
    
    def decode_char(self, ch):
        value = -1
        if ch >= 65 and ch <= 90:
            value = ch - 65
        elif ch >= 97 and ch <= 122:
            value = ch - 97 + 26
        elif ch >= 48 and ch <= 57:
            value = ch - 48 + 52
        elif ch == 43 or ch == 45:
            value = 62
        elif ch == 47 or ch == 95:
            value = 63
        elif ch == 61:
            value = 0
        else:
            pass
        return value

class TestClass:
    def test_method(self):
        src1 = [4, 5, 6]
        decoder = XorByte(0xff)
        dst1 = decoder.decode(src1)
        for x in dst1:
            print x
        
        print "|"
        
        src2 = [97, 71, 86, 115, 98, 71, 56, 61]
        decoder2 = Base64()
        dst2 = decoder2.decode(src2)
        for x in dst2:
            print x

try:
    TestClass().test_method()
except Exception as err:
    print "Exception: " + err.message

File: CustomDecoder/results/CustomDecoder.ts
interface ICustomDecoder {
  decode(src: OneArray);
}

class XorByte implements ICustomDecoder {
  xorValue: number;

  constructor(xorValue: number) {
      this.xorValue = xorValue;
  }

  decode(src: OneArray) {
    const dest = [];
    
    for (let i = 0; i < src.length; i++) {
        dest.push(src[i] ^ this.xorValue);
    }
    
    return dest;
  }
}

class Base64 implements ICustomDecoder {
  decode(src: OneArray) {
    const dest = [];
    
    // 4 base64 chars => 3 bytes
    for (let i = 0; i < src.length; i += 4) {
        const ch0 = this.decodeChar(src[i]);
        const ch1 = this.decodeChar(src[i + 1]);
        const ch2 = this.decodeChar(src[i + 2]);
        const ch3 = this.decodeChar(src[i + 3]);
        
        const trinity = (ch0 << 18) + (ch1 << 12) + (ch2 << 6) + (ch3);
        
        dest.push(trinity >> 16);
        dest.push((trinity >> 8) & 0xff);
        dest.push(trinity & 0xff);
    }
    
    return dest;
  }
  
  decodeChar(ch: number) {
    let value = -1;
    if (ch >= 65 && ch <= 90) {
        // `A-Z` => 0..25
        value = ch - 65;
    } else if (ch >= 97 && ch <= 122) {
        // `a-z` => 26..51
        value = ch - 97 + 26;
    } else if (ch >= 48 && ch <= 57) {
        // `0-9` => 52..61
        value = ch - 48 + 52;
    } else if (ch == 43 || ch == 45) {
        // `+` or `-` => 62
        value = 62;
    } else if (ch == 47 || ch == 95) {
        // `/` or `_` => 63
        value = 63;
    } else if (ch == 61) {
        // `=` => padding
        value = 0;
    } else {
    }
    return value;
  }
}

class TestClass {
  testMethod() {
    const src1 = [4, 5, 6];
    const decoder = new XorByte(0xff);
    const dst1 = decoder.decode(src1);
    for (const x of dst1) {
        console.log(x);
    }
    
    console.log("|");
    
    const src2 = [97, 71, 86, 115, 98, 71, 56, 61];
    const decoder2 = new Base64();
    const dst2 = decoder2.decode(src2);
    for (const x of dst2) {
        console.log(x);
    }
  }
}

try {
  new TestClass().testMethod();
} catch(e) {
  console.log(`Exception: ${e.message}`);
}